src/renderer/components/+extensions/extensions.tsx

```diff
@@ -35,458 +57,457 @@ interface InstallRequestPreloaded extends InstallRequest {
+          dispose();
+        }
       }
-    }
+    );
   }
+}
 
-  async unpackExtension({ fileName, tempFile, manifest: { name, version } }: InstallRequestValidated) {
-    const displayName = extensionDisplayName(name, version);
-    const extensionId = path.join(extensionDiscovery.nodeModulesPath, name, "package.json");
+async function requestInstalls(filePaths: string[]): Promise<void> {
+  const promises: Promise<void>[] = [];
 
-    logger.info(`Unpacking extension ${displayName}`, { fileName, tempFile });
+  for (const filePath of filePaths) {
+    promises.push(requestInstall({
+      fileName: path.basename(filePath),
+      filePath,
+    }));
+  }
 
-    this.extensionStateStore.extensionState.set(extensionId, {
-      state: "installing",
-      displayName
-    });
-    this.startingInstall = false;
+  await Promise.allSettled(promises);
+}
 
-    const extensionFolder = this.getExtensionDestFolder(name);
-    const unpackingTempFolder = path.join(path.dirname(tempFile), `${path.basename(tempFile)}-unpacked`);
+async function installOnDrop(files: File[]) {
+  logger.info("Install from D&D");
+  await requestInstalls(files.map(({ path }) => path));
+}
 
-    logger.info(`Unpacking extension ${displayName}`, { fileName, tempFile });
+async function installFromUrlOrPath(installPath: string) {
+  const fileName = path.basename(installPath);
+  let disposer: Disposer;
 
-    try {
-      // extract to temp folder first
-      await fse.remove(unpackingTempFolder).catch(Function);
-      await fse.ensureDir(unpackingTempFolder);
-      await extractTar(tempFile, { cwd: unpackingTempFolder });
+  try {
+    // install via url
+    // fixme: improve error messages for non-tar-file URLs
+    if (InputValidators.isUrl.validate(installPath)) {
+      disposer = ExtensionInstallationStateStore.startPreInstall();
+      const { promise: filePromise } = downloadFile({ url: installPath, timeout: 60000 /*1m*/ });
+      const data = await filePromise;
 
-      // move contents to extensions folder
-      const unpackedFiles = await fse.readdir(unpackingTempFolder);
-      let unpackedRootFolder = unpackingTempFolder;
+      await requestInstall({ fileName, data }, disposer);
+    }
+    // otherwise installing from system path
+    else if (InputValidators.isPath.validate(installPath)) {
+      await requestInstall({ fileName, filePath: installPath });
+    }
+  } catch (error) {
+    const message = getMessageFromError(error);
 
-      if (unpackedFiles.length === 1) {
-        // check if %extension.tgz was packed with single top folder,
-        // e.g. "npm pack %ext_name" downloads file with "package" root folder within tarball
-        unpackedRootFolder = path.join(unpackingTempFolder, unpackedFiles[0]);
-      }
+    logger.info(`[EXTENSION-INSTALL]: installation has failed: ${message}`, { error, installPath });
+    Notifications.error(<p>Installation has failed: <b>{message}</b></p>);
+  } finally {
+    disposer?.();
```

Comment:
doesn't `requestInstall()` call this in the case where `disposer` is not null?

Reply:
The disposer that I created can be called multiple times safely.

Reply:
Could you have added the removal of `validatedRequest.tempFile` to the disposer?

Reply:
I tried that and I though it was too complicated. I tried it again and it isn't.